### PR TITLE
E2-1891: cache interface

### PR DESCRIPTION
**Pull request by gg4race**
What              | Where/Who
------------------|----------------------------------------
JIRA Issue        | [JIRA_TICKET_ID](https://leanplum.atlassian.net/browse/JIRA_TICKET_ID)
People Involved   | @colleague1, @colleague2

[Notes on PR descriptions](https://leanplum.atlassian.net/wiki/spaces/PR/pages/288424408/Creating+a+GitHub+Pull+Request)

### CODE REVIEW

File: Leanplum/Common/LPConstants.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #define LEANPLUM_API_VERSION @"1.0.6"
 #define LEANPLUM_CLIENT @"ios"
 #define LEANPLUM_SUPPORTED_ENCODING @"gzip"
+#define LP_IV @"__l3anplum_iv__"
```
Comment (e7mac) [inline Leanplum/Common/LPConstants.h]: do we need this to be so obfuscated?

File: Leanplum/Cache/LPCache.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+NS_ASSUME_NONNULL_BEGIN
+@class LPVariable;
+@class LPRegion;
+@class LPMessage;
+@class LPVariant;
+@class LPUpdateRule;
+@class LPEventRule;
+@class LPVariantDebugInfo;
+
+@interface LPCache : NSObject
```
Comment (e7mac) [inline Leanplum/Cache/LPCache.h]: can we make this a protocol and then have this class implement it?
That will help us change caching backings if needed

File: Leanplum/Common/LPConstants.h (hunk 1 above)
Comment (gg4race) [reply Leanplum/Common/LPConstants.h]: I think this should be kept the same as the old SDK in cases where something was encrypted using this

File: Leanplum/Protocol/LPCacheProtocol.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+@class LPVariable;
+@class LPRegion;
+@class LPMessage;
+@class LPVariant;
+@class LPUpdateRule;
+@class LPEventRule;
+@class LPVariantDebugInfo;
+
+@protocol LPCacheProtocol <NSObject>
```
Comment (e7mac) [inline Leanplum/Protocol/LPCacheProtocol.h]: in obj-c we typically use the "..ing" naming convention.

https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingBasics.html#//apple_ref/doc/uid/20001281-1002242-BBCIJGDB

Let's call this `LPCaching`

File: Leanplum/Cache/LPCache.h [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//  Created by Grace on 7/23/19.
+//  Copyright © 2019 Leanplum. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "LPCacheProtocol.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface LPCache : NSObject<LPCacheProtocol>
```
Comment (e7mac) [inline Leanplum/Cache/LPCache.h]: We can temporarily name it `LPCache`, but I envision this being called `LPSQLiteCache` or `LPUserDefaultsCache`, as the case may be.

File: Leanplum/Cache/LPCache.m [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+- (NSData *)encodeIntoBinary:(NSArray *)data {
+    return [NSData new];
+}
+
+- (NSArray *)decodeFromBinary:(NSData *)data {
+    return [NSArray new];
+}
+
+- (void)persist:(NSData *)data forKey:(NSString *)key {
+
```
Comment (e7mac) [inline Leanplum/Cache/LPCache.m]: adding `//TODO:` here and in other places that are incomplete, will help this PR be more self contained.

**Comment by gg4race** [comment]
i will make pr later today

**Pull request merged**